### ISSUE TITLE
[mqtt] restoreOnStartUp not working

### ISSUE THREAD

**Issue by lsiepel** [openhab/openhab2-addons#5879]
**Expected Behavior**
When peristence with restoreOnStartUp is set, i would expect that the stored vaue is actually set to the MQTT channel/bound item.

**Current Behavior**
restoreOnStartup is updating the channel and within a split second it is flipped back to the UNDEF state.

**Steps to Reproduce**
1. Create a MQTT bridge:
```
Bridge mqtt:broker:lplxvibroker "Bridge Label" [host="192.168.1.10", port=1883, secure="OFF",retainMessages=false] {
    Thing topic slimmemeter "MQTT Slimmemeter" {
        Channels:
        Type number : stroom_in_stand_laag      "Meterstand in laag"    [ stateTopic="/lplxvi/meterkast/stroom/in/laag" ]
	}
}
```
2. Create item:
```
`Number	verd0_meterkast_stroom		"Meterstand in laagtarief"		    <energy>	{ channel="mqtt:topic:lplxvibroker:slimmemeter:stroom_in_stand_laag"}
```
3. Setup persistence:
```
Items {
	verd0_meterkast_stroom	: strategy = restoreOnStartup, everyChange 
}
```
4. Start openHAB and after bridge is initialized  post value to topic with mosquito or whatever tool you use (retainMessages = false) and confirm it being stored in the persistence service (i used MySQL)
5. Restart openHAB  and observe the item value not restoring. The event.log would contain something like:
```
2019-08-02 12:25:10.078 [hingStatusInfoChangedEvent] - 'mqtt:broker:lplxvibroker' changed from UNINITIALIZED to INITIALIZING
2019-08-02 12:25:10.094 [hingStatusInfoChangedEvent] - 'mqtt:broker:lplxvibroker' changed from INITIALIZING to OFFLINE
2019-08-02 12:25:10.414 [vent.ItemStateChangedEvent] - verd0_meterkast_stroom changed from NULL to 9.0  _(I GUESS THIS IS FROM restoreOnStartUp )_
2019-08-02 12:25:10.441 [hingStatusInfoChangedEvent] - 'mqtt:broker:lplxvibroker' changed from OFFLINE to ONLINE
2019-08-02 12:25:10.447 [me.event.ThingUpdatedEvent] - Thing 'mqtt:broker:lplxvibroker' has been updated.
2019-08-02 12:25:10.459 [hingStatusInfoChangedEvent] - 'mqtt:topic:lplxvibroker:slimmemeter' changed from UNINITIALIZED to INITIALIZING
2019-08-02 12:25:10.475 [hingStatusInfoChangedEvent] - 'mqtt:topic:lplxvibroker:slimmemeter' changed from INITIALIZING to ONLINE
2019-08-02 12:25:10.480 [vent.ItemStateChangedEvent] - verd0_meterkast_stroom changed from 9.0 to UNDEF
```
It looks like the onStartUpRestore is called before the channel is initialized and when the channel is initialized it resets to UNDEF instead of setting the restoreOnStartUp  value. 
If that is true, i qould suggest one of two fixes:
1. Somehow postpone restoreOnStartUp  util channel is initialized 
2. Somehow keep reference to restoreOnStartUp value and set it after channel is initialzed.

**Context**
My smart meter publishes changed value's to the MQTT topic. Sometimes it takes 2-3 days to get a new value, retainedMessages=false. When the value is not available, the UI is incomplete and keeping track off statistics (with rules) is hardly possible.

**Workaround**
1. Setting the broker to retain messages. Can cause side effects and not always possible. 
2. Add a rule that is triggered at startup with a 1 minute timmer that gets the value from the persistence service and use postUpdate to set it to the channel.
(i hate workarounds :-) they always lead to new problems and complexity and you tend to forget them until you are troubleshooting another problem for x hours)

**Environment**
openHAB build #1651

**Comment by Rossko57** [comment]
I view the problem differently here.
Restore-on-startup is working exactly as it should - it's right there in the log, changing the Item from its initial NULL state to 9.0, and doing that before bindings and rules are running.  Good.

Problem area is indeed when the channel initializes and updates the Item.
That in itself is a normal process - but we would expect valid data to overwrite the (restored) Item state, once it becomes available.
The unexpected part is the immediate UNDEF value.

I think the desired behaviour is for binding/channel not to update the Item until an actual MQTT topic arrives (in the absence of "retain" available at broker).
That would automatically preserve restored value of course.

**Comment by lsiepel** [comment]
I think we say the same in different words :-)

**Comment by davidgraeff** [comment]
Hm. I had this conversation while implementing the binding. For the mqtt binding itself UNDEF is actually correct. Because the binding subscribes to the broker topic and receives... nothing. I would say using retained messages is how the inventors of the mqtt protocol saw this problem solved.

If "Restore-on-startup" would use a "command" instead of just updating the channels, the mqtt binding would publish that value to the broker and would also apply that value. To be honest, I think an option for "Restore-on-startup" is required here and nothing on the bindings side.

**Comment by lsiepel** [comment]
>  I would say using retained messages is how the inventors of the mqtt protocol saw this problem solved.
Isn't it that only the Publisher can set the retain flag? Because the publishing device is not alwaus under control of the openHAB user, there are situations that you can't set the retain flag. 

> Because the binding subscribes to the broker topic and receives... nothing.
I would assume the binding would do... nothing. :-) How do other bindings handle these situations?

**Comment by davidgraeff** [comment]
There are not many generic 2.x bindings actually. knx is the only other one that I can think of.

**Comment by Rossko57** [comment]
I appreciate that this is something of a philosophical question :)

Setting aside 'retain' for now, the root of the issue is that we're dealing with one-off **events** arriving from MQTT which get stored as item **states**.
There's an obvious logic that says no event, no state change.

It does seem reasonable to expect that an Item set to state X will remain that way until something happens.   When an MQTT message does arrive, we act on it and update Item.   During the time between messages, nothing happens.  All good.

Looking more closely at the start up process - we subscribe to some/mqtt/topic.  Almost certainly this happens between messages, that is, in the time where we would normally do nothing.  Why on this occasion would updating with UNDEF be a good idea?

You can argue that UNDEF will show that the Item state is not up to date.
But the binding cannot actually know this - there might be channels linked to other inputs, or as in this case the user has chosen to restore the state, or a rule has set some special state.  These are all user choices that should not be trampled over.

In the simplest case, there is one channel and no restore, and the state is NULL until the first message arrives (if ever).  It's already evident we have no valid data from MQTT yet.

In summary, the fact that there is no message for us right at this instant is the usual condition in MQTT, and I do not think we need to flag that in any special way.  This is not indicative of any failure or problem, it's normal.

**Comment by davidgraeff** [comment]
You can restart the mqtt broker as well during an openHAB restart. So all mqtt clients will assume after a reconnect, that an mqtt topic is "not-set" (eg UNDEF in openHAB speech).

You as a user cannot overwrite what the binding knows about a channel, which is exactly that, that the mqtt topic has an unknown state. Even if we imagine that I disable an initial "undef" message, as soon as the mqtt connection got flaky and required a reconnection, the binding will set a channel to "undef" again. You would totally run out of sync with other mqtt clients if you assume values for topics that could be totally different (by just not listing for a second etc).

 The mqtt protocol is a pub/sub protocol, not stateful (unless "retain" is set). So if you are not happy with the stateless behavior you really need a different protocol with set/get semantics like CoAP or re-program your sensors to post retained messages.

The difference to the knx protocol, now that I think about it, is that knx although pub/sub as well also allows get/set semantics. So the knx addon will probably set channels to undef and request values at startup.

**Comment by lsiepel** [comment]
I don't need other protocols. MQTT is fine and it would be great if it would support restoreOnStartup. If you don't think it is a good addition, then i'll just give up on this.

I thought about a ugly workaround within openHAB:  a Rule that respond to the channel going from a value state to UNDEF and doing a postUpdate to the previous value state. For the use case it would do the same.

**Comment by J-N-K** [comment]
I agree with @Rossko57. Since `REFRESH` commands are issued not very often, why not make that configurable?

**Comment by Rossko57** [comment]
@lsiepel 
> I thought about a ugly workaround within openHAB: a Rule that respond to the channel going from a value state to UNDEF and doing a postUpdate to the previous value state. 

I think you should try that.  So far as I can see, that's essentially what you have already tried with restore and with sending a command.  I thought, but realise now may be mistaken, that when you tried postUpdate your Items that promptly got overwritten with UNDEF. I've no MQTT to play with this.

@davidgraeff 

> You as a user cannot overwrite what the binding knows about a channel,

Absolutely.  But Item is not channel.   May I link an Item to two MQTT channels, let's say looking for infrequent alarms from two sources, how should that behave? If not, why not?

FWIW I don't have any quibbles with binding setting UNDEF in the event of say broker connection failure.

Zwave 2.x does not update Items in the absence of messages, sleeping battery devices may not check in until tomorrow.

**Comment by lsiepel** [comment]
postUpdate worked (just tested). sendCommand does not work because the channel is read only, it resets the item value. 

```
rule "workaround"
when
    Item verd0_meterkast_stroom_in_laag changed to UNDEF
then
        verd0_meterkast_stroom_in_laag.postUpdate(verd0_meterkast_stroom_in_laag.previousState().state)
end
```

**Comment by Rossko57** [comment]
@lsiepel Okay, that's good. I was under a misapprehension that channel was overriding state changes, all wrong.
Appreciate it's only an experiment, but careful with previousState as the database could be recording these active changes in parallel.  I think it'll be fine as UNDEF is not persisted.

> sendCommand does not work because the channel is read only.

This is a problem.  As we've done to death elsewhere, commands should be ignored by channel and not provoke an UNDEF state (unless maybe the command is REFRESH)
Could you also try this with `autoupdate="false"` for your Item? (We don't expect any state change at all for that)

**Comment by davidgraeff** [comment]
> Zwave 2.x does not update Items in the absence of messages, sleeping battery devices may not check in until tomorrow.

And what does happen if your openHAB or zwvae controller restarts in the mean time? What does other zwave listers see that joined the network later on? It will be "undef" / "unknown state". 

The binding doc is just very clear about stateful addon channels, that they need to reflect the known or unknown state at any time. That's why there is a second type of channel, the "trigger" channel, that is stateless and represents more accurately that mqtt actually only propagates short events (unless used with retained messages).

It is already a compromise and actually violates openHAB binding developer guidelines, that an mqtt binding channel keeps the last known topic state. A binding should actually never keep state, that's what *Items* are for in openHAB. A "REFRESH" should in theory always re-request the device or service in the background which is the mqtt broker, and that would return "UNDEF" if re-requested for a non retained topic.

You might understand the dilemma now. I just don't feel comfortable to introduce even more special cases and violate more binding developer guidelines.

**Comment by J-N-K** [comment]
If REFRESH is not supported because of the nature of the attached equipment, the binding could also ignore that command.

**Comment by davidgraeff** [comment]
> If REFRESH is not supported because of the nature of the attached equipment, the binding could also ignore that command.

That might be the right option for mqtt binding channels on non retained topics.

**Comment by Rossko57** [comment]
> And what does happen if your openHAB or zwvae controller restarts in the mean time? 

In short, nothing. zwave binding does not push UNDEF. Unrestored Items will stay NULL. That does not mean it's correct, just what it does. Binding does try to poll devices, replies (and so real updates) can trickle in over a day.

> A binding should actually never keep state, that's what Items are for in openHAB. A "REFRESH" should in theory always re-request the device or service in the background which is the mqtt broker, and that would return "UNDEF" if re-requested for a non retained topic.

I don't have any disputes with that.
The issue is about force feeding UNDEF to Item in absence of any particular device related happening.
I suppose the differing view is really about whether you should flag lack of meaningful information at binding start up.  For non-retained MQTT, we never have any.
As it's only about start up, Items with no restore or other channels will be NULL, there's no problem updating to UNDEF then, but neither is there any benefit.  A restored Item, or one updated by some other channel, gets trampled.

Maybe we should consider what restoreOnStartup is _for_.  I don't know if there's an official reason, but the way people actually use it is to have best available values until something new comes along, which is what leads us here.    Perhaps that's all wrong?

I appreciate that broker retain is a general approach - but in the limited case of openHAB we are most likely restarting the broker as well, so that path is useless (at startup) for many users.

Well, not my decision and I cannot make a case any clearer. Done :)

**Comment by rkoshak** [comment]
I've been doing some looking int MQTT prettly closely over the past few months and IMHO, the current behavior of the binding is correct from the MQTT perspective. If a message represents a persistent state (e.g. a light is ON) then it should be published with the retained flag set. I'm pretty certain this can be configured in both Tasmota and ESPEasy (need to verify). 

So the real problem arises because we are coding and/or using devices coded by others that are not properly using the MQTT protocol. That does still leave us with needing to deal with the problem of restoreOnStartup, but it is clear to me that the first approach to solve the problem should be to modify the device to properly use the retained flag (and turning off restoreOnStartup for those Items). If and only if that is impossible is a work around necessary.

Philosophically, whose responsibility is it to deal with such work arounds? Should it be handled in the binding or should it be handled by the users in Rules? I can see good arguments either way, but I think it's important to be clear here that the entire problem exists because of the incorrect implementation of the MQTT devices in the first place. What ever the solution becomes, I would hate for the MQTT binding to break the proper usage (i.e. retained messages).

**Comment by lsiepel** [comment]
It should work the same as the z-wave binding. If no message is posted to the topic, The binding should not send a UNDEF.

And i disagree that the devices do not properly use the protocol. They do use the protocol how it was designed. Setting the retained flag is a choice and some manufacturers choose not to use it, some do.

As a user i might have a different point of view then the manufacturer and use restoreOnStartup. Or even if i agree with the manufacturer i might chain multiple channels to an item, that is also causing trouble with the UNDEF-while-no-message-was-posted

**Comment by J-N-K** [comment]
I agree with @lsiepel (as I said before). If a binding inherently does not support refresh (what is the case here), it should not handle `REFRESH`.

**Comment by rkoshak** [comment]
MQTT was designed such that the retained flag be set to true when publishing state messages. Failing to do so is not using the protocol as it was designed. If a message represents a state, the protocol is designed such that it is sent as a retained message with a time to live. 

From the Mosquitto man page: (sited by the mqtt wiki)

> If a topic is only updated infrequently, then without a retained message, a newly subscribed client may have to wait a long time to receive an update. With a retained message, the client will receive an instant update.

The Homie convention defaults all messages as retained and only defines a few exceptions. (https://github.com/homieiot/convention/blob/develop/convention.md#properties)

> A non-retained property would be useful for momentary events (door bell pressed).

The use of retained in this way flows from the MQTT  design principles (https://github.com/mqtt/mqtt.github.io/wiki/Design-Principles) in particular 3, 5, and part of 7:

> 3. Zero administration (or as close as possible). Behave sensibly in response to unexpected actions and enable applications to "just work".

I.e., Don't require clients to implement something like restoreOnStartup to know the last known good state of a sensor.

> 5. Expect and cater for frequent network disruption (for low bandwidth, high latency, unreliable, high cost-to-run networks)

I.e. don't expect that all the clients that care about a sensor reading will be connected to the broker when you send the message.

> 7. Expect that client applications may have very limited processing resources available.

I.e. don't publish messages if nothing has changed just to make sure that all your clients get update if they happened to be offline last time you published. It wastes the client's resources.

Quotes from various tutorials about MQTT and the retained flag:

mntolia.com
> Use retained message to update the status of a device on a topic. For example:
> “device1/status” can be Offline/Online
> or
> “kitchenlights/status” can be “On/Off”.

hivemq.com
>A retained message makes sense when you want newly-connected subscribers to receive messages immediately (without waiting until a publishing client sends the next message). This is extremely helpful for status updates of components or devices on individual topics. For example, the status of device1 is on the topic myhome/devices/device1/status. When retained messages are used, new subscribers to the topic get the status (online/offline) of the device immediately after they subscribe. The same is true for clients that send data in intervals, temperature, GPS coordinates, and other data. Without retained messages, new subscribers are kept in the dark between publish intervals. Using retained messages helps provide the last good value to a connecting client immediately.

thenewstack.io

>For example, when a mobile application (publisher) sends a message with the brightness and color values to multiple connected bulbs (subscriber), it expects that the new bulbs are set to the same values. So, when a new bulb connects to the MQTT broker, it receives the last message that all other bulbs would have received. This ensures that the state of existing subscribers and new subscribers is the same.

pauper.es
>[MQTT Retained messages are f]or IoT use cases where asynchronous operations are required where sensors (for example) are dormant most of the time and send data to the topics without knowing if there’s anyone listening to the topic. ...
>A simple scenario might be:
>
>Consumer subscribes to the topic /home/temperature and finds nothing and goes to sleep for 5 minutes;
>Publisher connects to the server and sends its temperature (21 degrees) to the topic /home/temperature and goes to sleep for 5 minutes;
>Consumer wakes up, subscribe to the topic /home/temperature and immediately, it gets the message 21.
>This is quite practical for scenarios both the publisher and subscriber are battery powered.
>
>Even when the subscriber is connected to the mains and battery is not an issue, it is great to regain the state of any given topic on re-connection after connection loss or power shortage.
>
>I cannot think of many IoT scenarios where both publishers and consumers want to keep wasting battery power for nothing.

I'm not just making this stuff up. Requiring the MQTT Client, in this case openHAB, to use it's own resources to guess at the last known good state of a sensor reading or device state is a violation of the very design principals that drove the creation of the protocol. How else can you define a use of something that goes against the very intent of that thing other than a misuse? It's wrong.  It's not just a mere choice. The manufactures who don't use retained for these types of messages (and when MQTT 5 get's broader support, time to live) are doing it wrong. They are using the MQTT protocol against it's very own design principals.

And yet, I still haven't had anyone give me actual examples of a firmware or script provided by a third party (i.e. not self developed) that does not provide at a minimum the option to have it publish stateful information using the retained bit. Is this Issue looking for a solution to a theoretical problem?

**Comment by davidgraeff** [comment]
Thanks for looking this up Rich. I guess that actually everybody agree. Some people just want to apply workarounds for their not-as-intended setup. There's no reason to prevent that if a change doesn't break other setups.

To my understanding, this issue boils down to: Either let the binding set channels to UNDEF (status quo) or let the core set channels to UNDEF implicitly (which allows "hacks" like restoreOnStartup). 

When I started to develop bindings, I thought it is always the responsibility of a binding to initialize all its channels, even if it just tells the framework that the state is "undefined". I thought that the framework would handle this in a smart way, so that it doesn't collide with restoreOnStartup.

The same for a "refresh" call from the framework: It it explicitly asks me about a channel, but I don't know anything about it yet, I answer with UNDEF. Without a response, how does the framework get to know that I received that request?

Now that I know OH better, I would sanctify a PR or make a PR own my own that removes those explicit UNDEFs for non-retained topic channels. Because: The framework doesn't need any response to a Refresh call (very rude API, if you ask me, but it is how it is). And the framework implicitly sets channels to UNDEF on start up.

**Comment by yozik04** [comment]
I think here is one more additional bug to fix. If a topic is retained in the brocker, openhab will not pick up the value during initialisation.

**Comment by davidgraeff** [comment]
This is more of a feature request. Open another issue to track it.
